src/image_surface.rs

```diff
@@ -44,6 +110,31 @@ impl ImageSurface {
         }
     }
 
+    #[cfg(feature = "png-functions")]
+    pub fn create_from_png_stream<R: Read>(stream: R) -> io::Result<ImageSurface> {
+        let mut env = png_functions::ReadEnv{ reader: stream, error: None };
+        let surface: ImageSurface = unsafe { from_glib_full(ffi::cairo_image_surface_create_from_png_stream(
+            transmute(png_functions::read_func::<R>), transmute(&mut env))) };
+        match env.error {
+            None => match surface.0.status() {   // The surface migth still be "nil" if the error occured in Cairo
+                Status::Success => Ok(surface),
+                Status::ReadError => Err(io::Error::new(io::ErrorKind::InvalidData, "Invalid PNG data.")),
+                Status::NoMemory => Err(io::Error::new(io::ErrorKind::InvalidData, "Could not allocate enough memory.")),
+                _ => Err(io::Error::new(io::ErrorKind::InvalidData, "Unknown error.")),
```

Comment:
Eh, they do seem to take some shortucts but then also return a number of statuses that make sense.